File: crates/hyper_rhi/src/vulkan/surface.rs
/*
 * Copyright (c) 2024, SkillerRaptor
 *
 * SPDX-License-Identifier: MIT
*/

use std::sync::atomic::Ordering;

use ash::{
    khr::{surface, swapchain},
    vk,
};
use raw_window_handle::{DisplayHandle, HasDisplayHandle, HasWindowHandle, WindowHandle};

use crate::{
    surface::SurfaceDescriptor,
    vulkan::{GraphicsDevice, Texture},
};

pub struct Surface {
    height: u32,
    width: u32,
    resized: bool,

    current_texture_index: u32,
    textures: Vec<Texture>,

    swapchain: vk::SwapchainKHR,
    swapchain_loader: swapchain::Device,
    inner: vk::SurfaceKHR,
    loader: surface::Instance,

    graphics_device: GraphicsDevice,
}

impl Surface {
    pub(crate) fn new(graphics_device: &GraphicsDevice, descriptor: &SurfaceDescriptor) -> Self {
        let loader = surface::Instance::new(graphics_device.entry(), graphics_device.instance());
        let surface = Self::create_surface(
            &graphics_device,
            &descriptor
                .window
                .display_handle()
                .expect("failed to get display handle"),
            &descriptor
                .window
                .window_handle()
                .expect("failed to get window handle"),
        );

        let size = descriptor.window.inner_size();
        let swapchain_loader =
            swapchain::Device::new(graphics_device.instance(), graphics_device.device());
        let swapchain = Self::create_swapchain(
            &graphics_device,
            size.x,
            size.y,
            &loader,
            surface,
            &swapchain_loader,
            vk::SwapchainKHR::null(),
        );

        let textures = Self::create_textures(
            graphics_device,
            &swapchain_loader,
            swapchain,
            size.x,
            size.y,
        );

        Self {
            height: size.y,
            width: size.x,
            resized: false,

            current_texture_index: 0,

            textures,
            swapchain,
            swapchain_loader,
            inner: surface,
            loader,

            graphics_device: graphics_device.clone(),
        }
    }

    fn create_surface(
        graphics_device: &GraphicsDevice,
        display_handle: &DisplayHandle<'_>,
        window_handle: &WindowHandle<'_>,
    ) -> vk::SurfaceKHR {
        let surface = unsafe {
            ash_window::create_surface(
                graphics_device.entry(),
                graphics_device.instance(),
                display_handle.as_raw(),
                window_handle.as_raw(),
                None,
            )
        }
        .expect("failed to create surface");

        surface
    }

    fn create_swapchain(
        graphics_device: &GraphicsDevice,
        width: u32,
        height: u32,
        surface_loader: &surface::Instance,
        surface: vk::SurfaceKHR,
        loader: &swapchain::Device,
        old_swapchain: vk::SwapchainKHR,
    ) -> vk::SwapchainKHR {
        let surface_capabilities = unsafe {
            surface_loader.get_physical_device_surface_capabilities(
                graphics_device.physical_device(),
                surface,
            )
        }
        .expect("failed to get surface capabilities");
        let surface_extent = Self::choose_extent(width, height, &surface_capabilities);

        let surface_formats = unsafe {
            surface_loader
                .get_physical_device_surface_formats(graphics_device.physical_device(), surface)
        }
        .expect("failed to get surface formats");
        let surface_format = Self::choose_surface_format(&surface_formats);

        let surface_present_modes = unsafe {
            surface_loader.get_physical_device_surface_present_modes(
                graphics_device.physical_device(),
                surface,
            )
        }
        .expect("failed to get surface present modes");
        let surface_present_mode = Self::choose_present_mode(&surface_present_modes);

        let max_image_count = surface_capabilities.max_image_count;
        let mut image_count = surface_capabilities.min_image_count + 1;
        if max_image_count > 0 && image_count > max_image_count {
            image_count = max_image_count;
        }

        let create_info = vk::SwapchainCreateInfoKHR::default()
            .surface(surface)
            .min_image_count(image_count)
            .image_format(surface_format.format)
            .image_color_space(surface_format.color_space)
            .image_extent(surface_extent)
            .image_array_layers(1)
            .image_usage(vk::ImageUsageFlags::COLOR_ATTACHMENT | vk::ImageUsageFlags::TRANSFER_DST)
            .image_sharing_mode(vk::SharingMode::EXCLUSIVE)
            .queue_family_indices(&[])
            .pre_transform(surface_capabilities.current_transform)
            .composite_alpha(vk::CompositeAlphaFlagsKHR::OPAQUE)
            .present_mode(surface_present_mode)
            .clipped(true)
            .old_swapchain(old_swapchain);

        let swapchain = unsafe { loader.create_swapchain(&create_info, None) }
            .expect("failed to create swapchain");

        swapchain
    }

    fn choose_extent(
        width: u32,
        height: u32,
        capabilities: &vk::SurfaceCapabilitiesKHR,
    ) -> vk::Extent2D {
        if capabilities.current_extent.width != u32::MAX
            || capabilities.current_extent.height != u32::MAX
        {
            return capabilities.current_extent;
        }

        vk::Extent2D {
            width: width.clamp(
                capabilities.min_image_extent.width,
                capabilities.max_image_extent.width,
            ),
            height: height.clamp(
                capabilities.min_image_extent.height,
                capabilities.max_image_extent.height,
            ),
        }
    }

    fn choose_surface_format(formats: &[vk::SurfaceFormatKHR]) -> vk::SurfaceFormatKHR {
        *formats
            .iter()
            .find(|&format| {
                format.format == vk::Format::B8G8R8A8_UNORM
                    && format.color_space == vk::ColorSpaceKHR::SRGB_NONLINEAR
            })
            .unwrap_or_else(|| &formats[0])
    }

    fn choose_present_mode(present_modes: &[vk::PresentModeKHR]) -> vk::PresentModeKHR {
        *present_modes
            .iter()
            .find(|&present_mode| *present_mode == vk::PresentModeKHR::MAILBOX)
            .unwrap_or(&vk::PresentModeKHR::FIFO)
    }

    fn create_textures(
        graphics_device: &GraphicsDevice,
        loader: &swapchain::Device,
        swapchain: vk::SwapchainKHR,
        width: u32,
        height: u32,
    ) -> Vec<Texture> {
        let images = unsafe { loader.get_swapchain_images(swapchain) }
            .expect("failed to get swapchain images");

        // TODO: Don't hardcode the format
        let textures = images
            .iter()
            .map(|&image| {
                Texture::new_external(
                    graphics_device,
                    image,
                    width,
                    height,
                    vk::Format::B8G8R8A8_UNORM,
                )
            })
            .collect::<Vec<_>>();

        textures
    }

    fn destroy_swapchain(&mut self) {
        unsafe {
            self.textures.clear();

            self.swapchain_loader
                .destroy_swapchain(self.swapchain, None);
        }
    }

    pub(crate) fn resize(&mut self, width: u32, height: u32) {
        self.height = height;
        self.width = width;
        self.resized = true;
    }

    pub(crate) fn current_texture(&mut self) -> Texture {
        // Resize
        if self.resized {
            let swapchain = Self::create_swapchain(
                &self.graphics_device,
                self.width,
                self.height,
                &self.loader,
                self.inner,
                &self.swapchain_loader,
                self.swapchain,
            );

            let textures = Self::create_textures(
                &self.graphics_device,
                &self.swapchain_loader,
                swapchain,
                self.width,
                self.height,
            );

            self.destroy_swapchain();

            self.swapchain = swapchain;
            self.textures = textures;

            self.resized = false;
        }

        let semaphores = [self.graphics_device.submit_semaphore()];
        let values = [self
            .graphics_device
            .current_frame_index()
            .load(Ordering::Relaxed) as u64
            - 1];
        let wait_info = vk::SemaphoreWaitInfo::default()
            .semaphores(&semaphores)
            .values(&values);

        unsafe {
            self.graphics_device
                .device()
                .wait_semaphores(&wait_info, u64::MAX)
                .expect("failed to wait for semaphore");
        }

        let (index, _) = unsafe {
            self.swapchain_loader.acquire_next_image(
                self.swapchain,
                u64::MAX,
                self.graphics_device.current_frame().present_semaphore,
                vk::Fence::null(),
            )
        }
        .expect("failed to acquire next image index");

        self.current_texture_index = index;

        self.textures[self.current_texture_index as usize].clone()
    }

    pub(crate) fn present(&mut self) {
        let swapchains = [self.swapchain];
        let wait_semaphores = [self.graphics_device.current_frame().render_semaphore];
        let image_indices = [self.current_texture_index];
        let present_info = vk::PresentInfoKHR::default()
            .swapchains(&swapchains)
            .wait_semaphores(&wait_semaphores)
            .image_indices(&image_indices);

        unsafe {
            self.swapchain_loader
                .queue_present(self.graphics_device.queue(), &present_info)
                .expect("failed to present swap chain");
        };

        let value = self
            .graphics_device
            .current_frame_index()
            .load(Ordering::Relaxed);
        self.graphics_device
            .current_frame_index()
            .store(value, Ordering::Relaxed);
    }

    pub(crate) fn loader(&self) -> &surface::Instance {
        &self.loader
    }

    pub(crate) fn inner(&self) -> vk::SurfaceKHR {
        self.inner
    }

    pub(crate) fn swapchain_loader(&self) -> &swapchain::Device {
        &self.swapchain_loader
    }

    pub(crate) fn swapchain(&self) -> vk::SwapchainKHR {
        self.swapchain
    }

    pub(crate) fn textures(&self) -> &[Texture] {
        &self.textures
    }
}

impl Drop for Surface {
    fn drop(&mut self) {
        self.destroy_swapchain();

        unsafe {
            self.loader.destroy_surface(self.inner, None);
        }
    }
}

File: hyper_rhi/src/hyper_rhi/vulkan/surface.cpp
/*
 * Copyright (c) 2024, SkillerRaptor
 *
 * SPDX-License-Identifier: MIT
 */

#include "hyper_rhi/vulkan/surface.hpp"

#include <algorithm>

#include <GLFW/glfw3.h>

namespace hyper_rhi
{
    VulkanSurface::VulkanSurface(VulkanGraphicsDevice &graphics_device, const SurfaceDescriptor &descriptor)
        : m_graphics_device(graphics_device)
        , m_surface(VK_NULL_HANDLE)
        , m_swapchain(VK_NULL_HANDLE)
    {
        this->create_surface(descriptor.window);
        this->create_swapchain(descriptor.window);

        // TODO: Retrieve swapchain images
    }

    VulkanSurface::~VulkanSurface()
    {
        vkDestroySwapchainKHR(m_graphics_device.device(), m_swapchain, nullptr);
        vkDestroySurfaceKHR(m_graphics_device.instance(), m_surface, nullptr);
    }

    void VulkanSurface::create_surface(const hyper_platform::Window &window)
    {
        HE_VK_CHECK(glfwCreateWindowSurface(m_graphics_device.instance(), window.native_window(), nullptr, &m_surface));
        HE_ASSERT(m_surface != VK_NULL_HANDLE);
    }

    void VulkanSurface::create_swapchain(const hyper_platform::Window &window)
    {
        VkSurfaceCapabilitiesKHR surface_capabilities = {};
        HE_VK_CHECK(vkGetPhysicalDeviceSurfaceCapabilitiesKHR(m_graphics_device.physical_device(), m_surface, &surface_capabilities));

        const VkExtent2D surface_extent = VulkanSurface::choose_extent(window.width(), window.height(), surface_capabilities);

        uint32_t format_count = 0;
        HE_VK_CHECK(vkGetPhysicalDeviceSurfaceFormatsKHR(m_graphics_device.physical_device(), m_surface, &format_count, nullptr));

        std::vector<VkSurfaceFormatKHR> formats(format_count);
        HE_VK_CHECK(vkGetPhysicalDeviceSurfaceFormatsKHR(m_graphics_device.physical_device(), m_surface, &format_count, formats.data()));

        const VkSurfaceFormatKHR surface_format = VulkanSurface::choose_format(formats);

        uint32_t present_mode_count = 0;
        HE_VK_CHECK(vkGetPhysicalDeviceSurfacePresentModesKHR(m_graphics_device.physical_device(), m_surface, &present_mode_count, nullptr));

        std::vector<VkPresentModeKHR> present_modes(present_mode_count);
        HE_VK_CHECK(vkGetPhysicalDeviceSurfacePresentModesKHR(
            m_graphics_device.physical_device(), m_surface, &present_mode_count, present_modes.data()));

        const VkPresentModeKHR surface_present_mode = VulkanSurface::choose_present_mode(present_modes);

        uint32_t image_count = surface_capabilities.minImageCount + 1;
        if (surface_capabilities.maxImageCount > 0 && image_count > surface_capabilities.maxImageCount)
        {
            image_count = surface_capabilities.maxImageCount;
        }

        const VkSwapchainCreateInfoKHR create_info = {
            .sType = VK_STRUCTURE_TYPE_SWAPCHAIN_CREATE_INFO_KHR,
            .pNext = nullptr,
            .flags = 0,
            .surface = m_surface,
            .minImageCount = image_count,
            .imageFormat = surface_format.format,
            .imageColorSpace = surface_format.colorSpace,
            .imageExtent = surface_extent,
            .imageArrayLayers = 1,
            .imageUsage = VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT | VK_IMAGE_USAGE_TRANSFER_DST_BIT,
            .imageSharingMode = VK_SHARING_MODE_EXCLUSIVE,
            .queueFamilyIndexCount = 0,
            .pQueueFamilyIndices = nullptr,
            .preTransform = surface_capabilities.currentTransform,
            .compositeAlpha = VK_COMPOSITE_ALPHA_OPAQUE_BIT_KHR,
            .presentMode = surface_present_mode,
            .clipped = true,
            .oldSwapchain = VK_NULL_HANDLE,
        };

        HE_VK_CHECK(vkCreateSwapchainKHR(m_graphics_device.device(), &create_info, nullptr, &m_swapchain));
        HE_ASSERT(m_swapchain != VK_NULL_HANDLE);
    }

    VkExtent2D VulkanSurface::choose_extent(const uint32_t width, const uint32_t height, const VkSurfaceCapabilitiesKHR &capabilities)
    {
        if (capabilities.currentExtent.width != std::numeric_limits<uint32_t>::max() ||
            capabilities.currentExtent.height != std::numeric_limits<uint32_t>::max())
        {
            return capabilities.currentExtent;
        }

        const VkExtent2D extent = {
            .width = std::clamp(width, capabilities.minImageExtent.width, capabilities.maxImageExtent.width),
            .height = std::clamp(height, capabilities.minImageExtent.height, capabilities.maxImageExtent.height),
        };

        return extent;
    }

    VkSurfaceFormatKHR VulkanSurface::choose_format(const std::vector<VkSurfaceFormatKHR> &formats)
    {
        for (const VkSurfaceFormatKHR &format : formats)
        {
            if (format.format == VK_FORMAT_B8G8R8A8_UNORM && format.colorSpace == VK_COLOR_SPACE_SRGB_NONLINEAR_KHR)
            {
                return format;
            }
        }

        return formats[0];
    }

    VkPresentModeKHR VulkanSurface::choose_present_mode(const std::vector<VkPresentModeKHR> &present_modes)
    {
        for (const VkPresentModeKHR &present_mode : present_modes)
        {
            if (present_mode == VK_PRESENT_MODE_MAILBOX_KHR)
            {
                return present_mode;
            }
        }

        return VK_PRESENT_MODE_FIFO_KHR;
    }
} // namespace hyper_rhi